### ISSUE TITLE
Uncontrolled recursion loop in  gettimeofday_sleep( ) 

### ISSUE THREAD

**Issue by Loginsoft-Research** [appneta/tcpreplay#549]
**Description** -As per our research on tcpreplay we found an infinite loop generating recursively at function gettimeofday_sleep( ) at file sleep.h which can lead to denial of service attack.

**Command** - : `tcpreplay-edit -r 80:84 -s 20 -b -C -m 1500 -K --no-flow-stats -P --stats=1 -i $INTERFACE $POC`
**POC** - [REPRODUCER](https://github.com/SegfaultMasters/covering360/blob/master/tcpreplay/TCPR_IF_POC?raw=true)

**Degub** -
```
Gdb:

: [ Legend: Modified register | Code | Heap | Stack | String ]
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── registers ────
$rax   : 0x0               
$rbx   : 0x00007fffffffdd10  →  0x000000005c8222c1
$rcx   : 0x00007ffff78b5827  →  <sched_yield+7> cmp rax, 0xfffffffffffff001
$rdx   : 0x151a            
$rsp   : 0x00007fffffffdc38  →  0x00000000004047c5  →  <tcpr_sleep+437> xor esi, esi
$rbp   : 0x5c8322bc        
$rsi   : 0x0               
$rdi   : 0x00007fffffffdbfc  →  0xffffdcf000000002
$rip   : 0x00007ffff78b5827  →  <sched_yield+7> cmp rax, 0xfffffffffffff001
$r8    : 0x2               
$r9    : 0x00007fffffffdd10  →  0x000000005c8222c1
$r10   : 0x00007fffffffdc40  →  0x0000000000000000
$r11   : 0x246             
$r12   : 0x000000000064e690  →  0x0000000100000001
$r13   : 0x52dde           
$r14   : 0x0000000000642010  →  0x00000000006429d0  →  0x000000000064d580  →  0x0000337330706e65 ("enp0s3"?)
$r15   : 0x4a              
$eflags: [carry PARITY adjust ZERO sign trap INTERRUPT direction overflow resume virtualx86 identification]
$cs: 0x0033 $ss: 0x002b $ds: 0x0000 $es: 0x0000 $fs: 0x0000 $gs: 0x0000 
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── stack ────
0x00007fffffffdc38│+0x0000: 0x00000000004047c5  →  <tcpr_sleep+437> xor esi, esi     ← $rsp
0x00007fffffffdc40│+0x0008: 0x0000000000000000     ← $r10
0x00007fffffffdc48│+0x0010: 0x00000000006429d0  →  0x000000000064d580  →  0x0000337330706e65 ("enp0s3"?)
0x00007fffffffdc50│+0x0018: 0x000000000064e690  →  0x0000000100000001
0x00007fffffffdc58│+0x0020: 0xccbef4846cea5600
0x00007fffffffdc60│+0x0028: 0x0000000000000002
0x00007fffffffdc68│+0x0030: 0x00000000006429d0  →  0x000000000064d580  →  0x0000337330706e65 ("enp0s3"?)
0x00007fffffffdc70│+0x0038: 0x000000000064e690  →  0x0000000100000001
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── code:x86:64 ────
   0x7ffff78b581b                  nop    DWORD PTR [rax+rax*1+0x0]
   0x7ffff78b5820 <sched_yield+0>  mov    eax, 0x18
   0x7ffff78b5825 <sched_yield+5>  syscall 
→ 0x7ffff78b5827 <sched_yield+7>  cmp    rax, 0xfffffffffffff001
   0x7ffff78b582d <sched_yield+13> jae    0x7ffff78b5830 <sched_yield+16>
   0x7ffff78b582f <sched_yield+15> ret    
   0x7ffff78b5830 <sched_yield+16> mov    rcx, QWORD PTR [rip+0x2d9641]        # 0x7ffff7b8ee78
   0x7ffff78b5837 <sched_yield+23> neg    eax
   0x7ffff78b5839 <sched_yield+25> mov    DWORD PTR fs:[rcx], eax
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── threads ────
[#0] Id 1, Name: "tcpreplay-edit", stopped, reason: SIGINT
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── trace ────
[#0] 0x7ffff78b5827 → sched_yield()
[#1] 0x4047c5 → gettimeofday_sleep(flush=0x0, now=0x7fffffffdd10, nap=0x642860, sp=0x64e690)
[#2] 0x4047c5 → tcpr_sleep(ctx=0x642010, sp=0x64e690, nap_this_time=0x642860, now=0x7fffffffdd10)
[#3] 0x405dc9 → send_packets(ctx=0x642010, pcap=0x0, idx=0x0)
[#4] 0x408b99 → replay_file(idx=0x0, ctx=0x642010)
[#5] 0x408b99 → tcpr_replay_index(ctx=0x642010)
[#6] 0x4083f8 → tcpreplay_replay(ctx=0x642010)
[#7] 0x402f91 → main(argc=0x1, argv=<optimized out>) 
```

**Comment by GabrielGanne** [comment]
Hi,

I'm willing to have a look at this, but do not quite understand the issue.

This is what I see in the capture file you provided: the second packet of the file is in the future.
```
# tshark -r TCPR_IF_POC
    1   0.000000    127.0.0.1 → 127.0.0.1    TCP 74 43230 → 389 [SYN] Seq=0 Win=32767 Len=0 MSS=16396 SACK_PERM=1 TSval=888227805 TSecr=0 WS=4
    2 65536.000028    127.0.0.1 → 127.0.0.1    TCP 74 389 → 43230 [SYN, ACK] Seq=0 Ack=1 Win=32767 Len=0 MSS=16396 SACK_PERM=1 TSval=888227805 TSecr=888227805
    3   0.000049 00:00:00_00:00:00 → 00:00:00_00:00:00 IPv4 66 Bogus IP header length (16, must be at least 20)
    4   0.000092    127.0.0.1 → 131.0.0.1    IPv4 80 Fragmented IP protocol (proto=TCP 6, off=64, ID=2f36)
    5   0.000104 00:00:00_20:00:00 → 00:00:00_00:00:00 0x08fa 66 Ethernet II

tshark: The file "TCPR_IF_POC" appears to have been cut short in the middle of a packet.
```

This means that `tcpreplay` or `tcpreplay-edit` will sleep for 65536s (45 hours) after sending the first packet.
The correction on the timestamps will only occur beginning on the third packet which appears to have been sent in the past (from the perspective of the second).
the program is "stuck" waiting, which is not the same as stuck in an infinitely recursive loop, which would crash (stack overflow) at some point.

Now, if you wish for `tcpreplay-edit` or `tcpreplay-edit` to control the speed at which packets are sent, you can do so by using the `-t` (topspeed), or `--pps` (packet-per-seconds) options.
I do not believe tcpreplay comes with a way of knowing that it is the **2nd** packet that is incorrect and not the **3rd**. There is no looking-ahead mechanism in place.

Does this analysis seem correct to you ?

Best regards,

**Comment by fklassen** [comment]
@GabrielGanne thanks for the analysis.

I agree that this is not a bug. As a tools and test program I consider it acceptable enter a `sched_yield()` loop while waiting for the next packet, even if the next packet is 45 hours later. That is one of the tricks to keep performance high, and since `sleep()` returns at an unpredictable time, it is our only option to keep the program usable.

I assure you that a `sched_yield()` loop is not a service denial attack. I've tested this on single-core machines, and the CPU remains very responsive. Other than seeing 100% CPU usage reported, you would hardly know that something is running in a tight loop.

Closing as "not a bug".
